Add device_id support to FSDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77321

Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.


Benchmarking on single host 8 GPUs:

- GPT 700m parameter model FSDP init on CPU: ~28s, moving to GPU speeds it up to 2s, a 14x win
- GPT 6B parameter model FSDP init on CPU: ~47s, moving it to GPU speeds it up to 17s, 2.7x win

rohan-varma committed:
device_id

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -660,10 +679,62 @@ def _run_param_init_fn():
 
+        # if device_id is specified, ensure it is the same
+        assert (
+            not self.device_id or self.compute_device == self.device_id
```
self.device_id is not None

rohan-varma committed:
Update on "[WIP] Add device_id support to FSDP"




[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Add device_id support to FSDP"




[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3519,7 +3627,7 @@ def _get_default_cuda_device(module: nn.Module) -> torch.device:
         pass
     # Fall back to current CUDA device
-    return torch.device("cuda")
+    return torch.device("cuda", torch.cuda.current_device())
```
@zhaojuanmao I think this change should be okay,  this makes it truly the current cuda device, instead of the device just being "cuda", it will be like "cuda:0" or "cuda:1", but can you confirm this is ok?

zhaojuanmao replied:
sounds good, thanks!

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -660,10 +680,66 @@ def _run_param_init_fn():
         self.compute_device = _get_default_cuda_device(module)
 
+        # if device_id is specified, ensure it is the same
+        assert (
```
Curious: I am guessing that this `assert` is meant as an internal `assert` and that you updated the `return` of `_get_default_cuda_device()` so to avoid the case where `torch.device("cuda") != torch.device("cuda:0")` even though they will be redirected to the same logical device?

awgu replied:
My concern is what if the _user_ passes in `torch.device("cuda")`? It correctly has type `Union[int, torch.device]`, so it follows the API. However, would this `assert` be triggered given the new `_get_default_cuda_device()` `return`? (E.g. if `self.compute_device` is `torch.device("cuda:0")`, but `self.device_id` is `torch.device("cuda")`.)

More broadly as well, does this API require the user to call `torch.cuda.set_device()`? There has been confusion about that method's usage (https://github.com/pytorch/pytorch/issues/50112), so if we recommend it as part of the FSDP workflow, we may want to clarify that somewhere.

rohan-varma replied:
The following is a bit unfortunate, though maybe I'm missing some context:

```
>>> import torch
>>> torch.cuda.set_device(0)
>>> z = torch.device("cuda")
>>> m = torch.device("cuda", torch.cuda.current_device())
>>> z == m
False
```

I think as guidance we can:

1) Encourage user to call torch.cuda.set_device() in general for FSDP / DDP training
2) if user passes in torch.device("cuda"), set `device_id` to `torch.device("cuda", torch.cuda.current_device())`, to ensure that the index is set?

awgu replied:
> The following is a bit unfortunate, though maybe I'm missing some context:

I am not sure if there is a good way to address this. I encountered similar issues in the past and was unable to come up with anything.

Nonetheless, the guidance you outlined sounds great to me!

awgu reviewed (commented):
The PR looks good overall. I left some comments with two main ones to look at: (1) L699 and (2) L739, both in `fully_sharded_data_parallel.py`.

rohan-varma commented on test/distributed/fsdp/test_fsdp_ignored_modules.py:
```diff
@@ -109,6 +109,7 @@ def test_ignored_modules_nested(self):
             wrapped_model.run_backward(loss)
             optim.step()
 
+    @skip_if_lt_x_gpu(2)
```
Yeah, previously this test didn't require GPU but now it does due to the FSDP changes

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -308,6 +308,7 @@ def setUp(self) -> None:
         # For all the tests here, we use a fake group
         self.process_group = DummyProcessGroup(rank=0, size=1)
 
+    @unittest.skipIf(torch.cuda.device_count() < 2, "Requires at least 2 GPUs")
```
skip_if_lt_x_gpu doesn't work for test classes that don't inherit from FSDPTest

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -660,10 +680,66 @@ def _run_param_init_fn():
+        if self.device_id is not None:
+            param = None
+            try:
+                param = next(module.parameters())
```
Great point!

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -660,10 +680,66 @@ def _run_param_init_fn():
+            # device_id argument is not specified
+            # If module is on CPU, move it to current device and log a warning.
+            try:
+                param = next(module.parameters())
```
fixed

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -548,6 +549,15 @@ class FullyShardedDataParallel(nn.Module):
 
+        device_id (Optional[Union[int, torch.device]]): An ``int`` or ``torch.device``
+            describing the CUDA device FSDP module should be moved to before
+            initialization such as sharding takes place. If this argument is not specified
```
such as computation takes place

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -660,10 +680,66 @@ def _run_param_init_fn():
+        # Move module to device specified. Note that this is done prior to
+        # setting compute_device to ensure that they align.
+        needs_move_back_to_cpu = False
+        if self.device_id is not None:
```
nit: move this logic into a helper function

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -753,6 +829,13 @@ def _run_param_init_fn():
+        if needs_move_back_to_cpu:
+            # _fsdp_wrapped_module took ownership of module and we deleted
+            # module, so move _fsdp_wrapped_module back.
+            self._fsdp_wrapped_module = self._fsdp_wrapped_module.to(
+                torch.device("cpu")
+            )
+
```
just want to make sure this is a blocking operation, right?

rohan-varma replied:
yep, `non_blocking=False` is the default

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -628,6 +639,15 @@ def __init__(
+                device_id if isinstance(device_id, torch.device)
+                else torch.device(device_id)
+            )
+            assert isinstance(self.device_id, torch.device)
```
removed

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1933,6 +2037,17 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
 
             # Start of a forward pass.
             self.training_state = TrainingState_.FORWARD
+            if self._is_root:
```
could we add a documentation to let users know this new support?

rohan-varma replied:
Added

zhaojuanmao reviewed (commented):
looks good to me overall, just some minor nits. thanks a lot for adding this!

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

rohan-varma committed:
Update on "Add device_id support to FSDP"


Adds `device_id` argument to FSDP. If this is specified, and module is on CPU, we move module to this device.

In addition, even if device_Id is not specified, we move module to this device to do flattening, sharding, etc but then move it back to CPU before returning to user.

This PR also moves module inputs to the `compute_device` during forward of root FSDP instance. Note that we guarantee `compute_device == device_id` if `device_id` is specified.

This PR also detects if user passed in multi-device module, which FSDP does not support, and throws an error.

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -429,8 +430,12 @@ class FullyShardedDataParallel(nn.Module):
         device is set properly using ``torch.cuda.set_device(device_id)``.
+        Alternatively, if module is input on CPU, ``device_id`` argument can
+        be specified to move module to CUDA device during initialization.
         FSDP will get compute device from module first, if module device
```
nit: Since these warnings are user-visible in the docs, I am more inclined to be picky about style/grammar. Since I am the one who is picky, I can make a pass over the user-visible documentation before branch cut.

E.g.
- Maybe we should backtick `module` (i.e. \`\`module\`\`) since it specifically refers to the `module` argument in the constructor.
- Some grammatical changes:

> If the destination CUDA device has ID `dev_id`, either (1) `module` should already be placed on that device, (2) the device should be set using `torch.cuda.set_device(dev_id)`, or (3) `dev_id` should be passed into the `device_id` constructor argument. This FSDP instance's compute device is CPU only if `module` is input on CPU and the `device_id` argument is not specified. Otherwise, the compute device is the CUDA device with ID `dev_id`.

rohan-varma replied:
sure, we can work together to make sure the user facing documentation is as clear as possible.

awgu reviewed (approved):
Looks good to me! Thanks for the quick turnaround.

rohan-varma commented:
@pytorchbot merge this please

awgu commented:
@agneet42 What exactly are you logging? When does the first value become `cuda:1`?

Are you calling `torch.cuda.set_device()` on each rank?
